File: go/ZigZagLevelOrder/ZigZagLevelOrder.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func zigzagLevelOrder(root *TreeNode) [][]int {
	queue := make([][]*TreeNode, 0)
	result := make([][]int, 0)
	queue = append(queue, []*TreeNode{root})
	dir := "left"

	for len(queue) != 0 {
		activeQueue := queue[0]
		queue = queue[1:]
		nextQueue := make([]*TreeNode, 0)
		currResult := make([]int, 0)

		for _, node := range activeQueue {
			if node == nil {
				continue
			}

			currResult = append(currResult, node.Val)

			nextQueue = append(nextQueue, node.Left)
			nextQueue = append(nextQueue, node.Right)
		}

		if len(currResult) > 0 {
			if dir == "right" {
				for i, j := 0, len(currResult)-1; i < j; i, j = i+1, j-1 {
					currResult[i], currResult[j] = currResult[j], currResult[i]
				}
				result = append(result, currResult)
				dir = "left"
			} else {
				result = append(result, currResult)
				dir = "right"
			}
		}

		if len(nextQueue) > 0 {
			queue = append(queue, nextQueue)
		}
	}
	return result
}

func main() {
	left := &TreeNode{Val: 1}
	right := &TreeNode{Val: 3}
	root := &TreeNode{Val: 2, Left: left, Right: right}
	fmt.Println(zigzagLevelOrder(root))
}

File: python/zigzagLevelOrder.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


def zigzagLevelOrder(root: TreeNode | None) -> list[list[int]]:
    if not root:
        return []

    queue = []
    result = []
    queue.append([root])
    dir = "left"
    while len(queue) != 0:
        activeQueue = queue.pop(0)
        currResult = []
        nextQueue = []
        for node in activeQueue:
            if node is None:
                continue

            currResult.append(node.val)

            nextQueue.append(node.left)
            nextQueue.append(node.right)

        if len(currResult) > 0:
            if dir == "right":
                i, j = 0, len(currResult) - 1
                while i < j:
                    currResult[i], currResult[j] = currResult[j], currResult[i]
                    i += 1
                    j -= 1

                dir = "left"
                result.append(currResult)
            else:
                dir = "right"
                result.append(currResult)

        if len(nextQueue) > 0:
            queue.append(nextQueue)

    return result


left = TreeNode(val=1)
right = TreeNode(val=3)
root = TreeNode(2, left, right)
print(zigzagLevelOrder(root))
